Rewrite status bar updater

kaste opened the pull request:
Fixes #1093
Fixes #1094 
Fixes #1095
Fixes #1102 

Introduce global app state at `core.state`. We now have a simple architecture: `core.repo_status` listens for some events and produces state updates (by calling and parsing `git status`), `core.status_bar_view` subscribes to the `state` (could be also called *store*) and update the status bars.

kaste commented:
Introduce global app state at `core.state`. We now have a simple architecture: `core.repo_status` listens for some events and produces state updates (by calling and parsing `git status`), `core.status_bar_view` subscribes to the `state` (could be also called *store*) and update the status bars. 

This already has the benefit if you're have multiple windows into the same repo. We go further by making some interfaces subscribe to the state as well. Then they will update automatically, and instantly, on updates. 

The state is generally useful for a lot of other commands. E.g. when we show the checkout panel, we want to show all branches minus the current one for which we call `git` to tell us right now. After this change we can just ask the `state` which branch we're currently on. E.g. when we `pull` we ask for the tracking/remote branch but this information is in the state object already now. E.g. when we `push`, we will know that this will fail if we're `behind`.

kaste commented on core/repo_status.py:
```diff
@@ -0,0 +1,249 @@
+        global active_view
+        active_view = view
+
+        maybe_update_status_for_view(view)
```
Thanks, that comment is outdated and must be removed. The functionality is now in `status_bar_view`

kaste commented on core/repo_status.py:
```diff
@@ -0,0 +1,249 @@
+def make_git(repo_path):
+    # type: (RepoPath) -> GitCommand
+    git = GitCommand()
+    git.repo_path = repo_path
```
We use the new setter `repo_path` here. It is not meant to be a general cache, bc we cache either on the `view.settings()` and of course in the new `repo_paths` global in `git_command`. What we need is a simple way to construct `GitCommand`s given a `repo_path` which is totally missing in the abstractions, super-classes. `make_git` itself should probably moved to `git_command`. Ideally, just `GitCommand(repo_path=...)` should be possible. 

Note that we currently store the `repo_path` in the `settings()` *only* for our fancy views because they otherwise have no `view.file_name()` anymore. After this PR, we don't use `settings` as a cache for *normal*, ordinary views the users uses to hack its code anymore. Normal views always get the value from the global cache because they have a `file_name()`. To the effect of that for the regular edge case of unsaved buffers we can still take the global cache.

kaste commented on core/state.py:
```diff
@@ -0,0 +1,88 @@
+def update_state(repo_path, partial_state):
+    # type: (RepoPath, RepoStatus) -> None
+    State[repo_path].update(partial_state)
+    notify_all(repo_path, State[repo_path])
```
TODO: `notify_all` MUST NOT run `@atomic`

kaste replied:
TODO: Use a simple `Lock` then

kaste commented on core/status_bar_view.py:
```diff
@@ -0,0 +1,55 @@
+        view.erase_status("gitsavvy-repo-status")
+
+
+subscribe('status_bar_updater', render_status)
```
Here is a `subscribe` on the module level (executed at import time). But where is the unsubscribe 😁?

In SublimeLinter land I have access to the `plugin_loaded`/`unloaded` pair which is the correct place to put such code. GitSavvy is not prepared for that. The dirty implementation here uses a key (here: `status_bar_updater`) to identify unique callbacks, so this implementation is safe for hot-reloading at least. It does not unload if the plugin gets uninstalled. We can get rid of the key if we had a mechanism to register functions as `loaded`/`unloaded` handlers because then the functions itself can be used as `id`. (The function `id` changes during hot-reload.)

kaste commented on core/state.py:
```diff
@@ -0,0 +1,88 @@
+            'unstaged': List[FileStatus],
+            'untracked': List[FileStatus],
+            'conflicts': List[FileStatus],
+            'short_status': str,
```
'staged/unstaged/untracked/conflicts' is derived state from 'file_statuses'. 'short_status' is derived from all the other properties. It is usually better to *not* store derived values in the main store.

kaste commented on core/repo_status.py:
```diff
@@ -0,0 +1,249 @@
+
+
+def format_short_status(info):
+    # type: (RepoStatus) -> str
```
'short_status' is derived state for the `status_bar_view`, the formatting can thus be moved to `status_bar_view`

kaste commented on core/repo_status.py:
```diff
@@ -0,0 +1,249 @@
+    pred = lambda: current_token.get(repo_path) == token
+    sublime.set_timeout_async(
+        partial(executor, pred, sink)
+    )
```
The implementation of `pred` and `executor` have changed a lot during this PR but the structure to have a `pred`, the `executor`, and the `sink` actually never changed. So I wouldn't do this. (Generally lambdas in python are hard to read except for the trivial cases. I usually avoid them because they're just not flexible enough to change.)

asfaltboy replied:
GitSavvy code is all full of lambdas on calls to `sublime.set_timeout_async` and similar functions that accept a callable.

Personally, lambda version reads fine to me, but of course defining a function that does the same is fine as well.

The executor, however, reads badly. One needs to follow the code a few times back and forth, at least the first time (just in case executor does something unexpected), as well as be familiar (or spend time looking up) some new concepts or abstractions (sink? thunk? pred? Thunk?) ... for such a short piece of code it seems like it's a shame to introduce all of these.

I think this sort of addition should be discussed (e.g with other maintainers) last we introduce too many paradigms for the project... mixed methods of code flow are just as confusing. It may well be that refactoring the whole project to use this sort of alternative "functional flow control" will result in much cleaner code, but then we should probably do that in a separate PR.

kaste replied:
You're correct that in the current state executor is 'refactored to an end', where it becomes ridiculous, strange or funny. This is still for changeability. Maybe we need a lock here. Everytime you hack on it, I add print messages to it e.g. to catch the `else: print('--> skipped updating'). 

So, *yes* the executor code became so short that I would write it probably different. *no* this is at the beginning of the lifecycle for the global app state. We should focus on exploring and changing.

kaste commented on core/git_command.py:
```diff
@@ -358,85 +426,31 @@ def git_binary_path(self):
+            raise RuntimeError("No active view exists.")
 
-        # fallback: use the first folder if the current file is not inside a git repo
+        repo_path = repo_path_for_view(view)
```
See comment below for `make_git`

kaste commented on core/git_command.py:
```diff
@@ -110,6 +110,74 @@ def communicate(self, stdin):
+    ).strip() or None
+
+
+def search_paths(view):
```
`search_paths` is probably not reusable code. It also currently introduces the miserable edge-case behavior for untracked projects, in that we first call git for the filepath, then for the project folder. But since it is an untracked project bot calls don't lead to a value, and we can't cache `None`s. So we do two toplevel calls over and over. (To the effect that we have a yanky experience for the untracked projects, basically when we ideally should consume no resources at all.)

kaste commented on common/util/debug.py:
```diff
@@ -54,6 +60,14 @@ def make_log_message(_type, **kwargs):
 
 def log_git(command, stdin, stdout, stderr, seconds):
     """ Add git command details to debug log """
+    global enabled
```
This is redundant but it is a `global`. (This is probably just a habit because then with the args and the one global I have all moving variables at sight.)

kaste commented:
Ah, `gs_update_status_bar` is still there just moved.

kaste commented:
I now attach the status dashboard to the store. This saves one 'status' call per 'activate', and just makes it super snappy. Look it even updates in the background 🚀 

![4Cwhv1ICIC](https://user-images.githubusercontent.com/8558/62387361-b0d5da80-b55a-11e9-8369-391be7d05187.gif)

Also fixes #1102.

asfaltboy commented:
Woah, this PR now has a lot of changes! We should freeze any additions to here, or we risk introducing regressions.

With that said, I'm super happy to test it out! The status dashboard does feel snappier and I love the automatic update.

There are a few unresolved discussions (both yours and mine), and I'm not sure how much more work is left on this feature. To me, it looks 90% done and if the branch is stable I'd like to merge it into `dev`.

Can you share if there is anything else you want to fix before we can merge?

kaste commented:
The only thing I don't like is the `subscribe/unsubscribe` story. Then, there is code duplication: bc I didn't want to risk regressions I had to copy and then modify functions  parsing the 'status' output. 

I think we have a lot of options from here:

- Go through other consumers of 'git status' and tell em to use the app state instead.
- Put e.g. 'git log -n1'  into the store bc it is used by other dashboards as well.
- Esp. make the branch dashboard use the same pattern as the status dashboard bc it looks like a rather simple view.
- Optimize commands. E.g. for 'git push' just use the branch and remote branch from the store instead of fetching it. E.g. for 'git rebase --continue' and 'git rebase --abort' we can enable/disable the command whether the repo state shows 'rebasing'. E.g. for 'git checkout' just use the active branch from the store. 
- Optimize invalidating/refreshing. E.g. after a checkout 'status' and 'log -n1' change but not 'stash list'. After a push 'status' changes but not 'log -n1' and 'stash list'. 
- Optimize 'on_activated'. If we just switch between tabs, very likely nothing can change (except we detect a Terminus terminal). 'on_activate' is more interesting if Sublime lost its focus and regains it. 'on_activate' between panel selections is 100% unneeded. (E.g. if the user invokes a command which in turn opens a panel. The view regains focus shortly exactly between `run` and `run_async`.)

kaste commented:
I converted the branch dashboard to the new snappy-style. I will not push it on top of this here, just saying. I could try the tags dashboard next. (This will probably get easier the more often you do it 😁 	)

This PR is okay. We can push fixup commits as long as we find something and then I make a nice logical rebase as a last step. 

I had a bullet list in my last comment: Anything you would really like to see?

asfaltboy commented:
Thanks for your work @kaste! The "branch dashboard" is something I really use very often, so I'm glad you tackled it first 👍 😃!

The first thing I would like to do is actually documenting this. Perhaps we can add a section about GitSavvy internals, and explain the reasoning and usage of the cached state mechanism.

Then, if nothing is broken, let's merge to dev and allow some testing time to everyone who uses our dev version (I know I do).

BTW: how come `util.disable_logging` was removed? Might be useful to run with logs and silent noisy parts? Or do you intend we use the "tagged" tracer option instead? P.s we should really document these things.

And, as for future improvements/features, I would think implementing `git log -n1` would be the next most reasonable step. I'd also like to see "remote refs" in the cached state. That said, everything on your list looks like great ideas. Perhaps we can make a separate ticket to maintain a list of TODOs and discussion there?

kaste commented:
I rebased and applied the numerous fixups in-place. I removed `util.disable_logging` simply because I removed the last usage of it. 🤷‍♂ 

I actually don't know how this all fits together. It sure will change when we also modernize the branch dashboard and a few other places. Today I just wanted the fancy graph dots. 😬

kaste commented:
I don't know how we should proceed here. It is also rather old so I don't have the details of this PR in my head.  I think this is not ready for dev. On the one hand it fixes many issues, on the other hand it introduced a lot of new stuff which only pays off if we proceed further in this direction. If we do not proceed it would leave a mess. Should I start an integration branch and then open further PRs on top of that?

stoivo commented:
I will go through it during this weekend.

kaste commented:
Closing.  Follow up PR is #1478.
